File: chifrementasymetric/fini/message/aleatoir.py
import random

def random_impair(min_val, max_val):
    n = random.randint(min_val, max_val)
    if n % 2 == 0:
        n += 1
        if n > max_val:
            n = min_val
    return n

def puissance_modulaire(a, d, n):
    res = 1
    a = a % n
    while d > 0:
        if d % 2 == 1:
            res = (res * a) % n
        d = d // 2
        a = (a * a) % n
    return res

def miller_rabin(n, k=5):
    if n <= 1:
        return False
    if n == 2 or n == 3:
        return True
    if n % 2 == 0:
        return False
    
    r = 0
    d = n - 1
    while d % 2 == 0:
        d //= 2
        r += 1
    
    for _ in range(k):
        a = random.randint(2, n - 2)
        x = puissance_modulaire(a, d, n)
        if x == 1 or x == n - 1:
            continue
        for _ in range(r - 1):
            x = (x * x) % n
            if x == n - 1:
                break
        else:
            return False
    return True

def generer_nombre_premier_chiffres(chiffres):
    if chiffres <= 0:
        return None
    debut = 10**(chiffres - 1)
    fin = 10**chiffres - 1
    while True:
        candidat = random_impair(debut, fin)
        if miller_rabin(candidat):
            return candidat

nombre_de_chiffres = 100
p = generer_nombre_premier_chiffres(nombre_de_chiffres)
q = generer_nombre_premier_chiffres(nombre_de_chiffres)

print(f"Nombre premier p avec {nombre_de_chiffres} chiffres : {p}")
print(f"Nombre premier q avec {nombre_de_chiffres} chiffres : {q}")

File: chifrementasymetric/fini/message/aleatoir.cpp
#include <iostream>
#include <ctime>

using namespace std;

// Paramètre : nombre de chiffres
int nombre_de_chiffres = 100;

// Renvoie un entier "aléatoire" basé sur clock()
unsigned long long random_seed() {
    return static_cast<unsigned long long>(clock());
}

// Modulo exponentiation : (base^exp) % mod
unsigned long long puissance_modulaire(unsigned long long base, unsigned long long exp, unsigned long long mod) {
    unsigned long long res = 1;
    base = base % mod;
    while (exp > 0) {
        if (exp % 2 == 1)
            res = (res * base) % mod;
        exp = exp / 2;
        base = (base * base) % mod;
    }
    return res;
}

// Test de primalité Miller-Rabin
bool miller_rabin(unsigned long long n, int k = 5) {
    if (n < 2)
        return false;
    if (n == 2 || n == 3)
        return true;
    if (n % 2 == 0)
        return false;

    // Trouver d et r tels que n-1 = 2^r * d
    unsigned long long d = n - 1;
    int r = 0;
    while (d % 2 == 0) {
        d /= 2;
        r++;
    }

    for (int i = 0; i < k; ++i) {
        unsigned long long a = 2 + random_seed() % (n - 3);
        unsigned long long x = puissance_modulaire(a, d, n);
        if (x == 1 || x == n - 1)
            continue;
        bool continue_outer = false;
        for (int j = 0; j < r - 1; ++j) {
            x = (x * x) % n;
            if (x == n - 1) {
                continue_outer = true;
                break;
            }
        }
        if (continue_outer)
            continue;
        return false;
    }
    return true;
}

// Génère un nombre impair aléatoire de n chiffres
unsigned long long random_impair(int chiffres) {
    unsigned long long debut = 1;
    for (int i = 1; i < chiffres; ++i)
        debut *= 10;
    unsigned long long fin = debut * 10 - 1;

    unsigned long long n = debut + (random_seed() % (fin - debut + 1));
    if (n % 2 == 0) n++; // Forcer impair
    if (n > fin) n = debut + 1; // Recycle si dépasse
    return n;
}

// Génère un nombre premier de n chiffres
unsigned long long generer_nombre_premier_chiffres(int chiffres) {
    while (true) {
        unsigned long long candidat = random_impair(chiffres);
        if (miller_rabin(candidat))
            return candidat;
    }
}

int main() {
    unsigned long long premier = generer_nombre_premier_chiffres(nombre_de_chiffres);
    cout << premier << endl;
    return 0;
}
